dot/network/discovery.go

```diff
@@ -0,0 +1,191 @@
+				time.Sleep(time.Second * 10) // wait for peers to connect before starting DHT
+			} else {
+				break
+			}
+
+			peers = d.h.Network().Peers()
+		}
+
+		for _, p := range peers {
+			d.bootnodes = append(d.bootnodes, d.h.Peerstore().PeerInfo(p))
+		}
+	}
+
+	logger.Info("starting DHT...", "bootnodes", d.bootnodes)
+
+	dhtOpts := []dual.Option{
+		dual.DHTOption(kaddht.Datastore(d.ds)),
+		dual.DHTOption(kaddht.BootstrapPeers(d.bootnodes...)),
+		dual.DHTOption(kaddht.V1ProtocolOverride(d.pid + "/kad")),
+		dual.DHTOption(kaddht.Mode(kaddht.ModeAutoServer)),
+		//dual.DHTOption(kaddht.Mode(kaddht.ModeServer)),
+	}
+
+	// create DHT service
+	dht, err := dual.New(d.ctx, d.h, dhtOpts...)
+	if err != nil {
+		return err
+	}
+
+	d.dht = dht
+	return d.discoverAndAdvertise()
+}
+
+func (d *discovery) stop() error {
+	if d.dht == nil {
+		return nil
+	}
+
+	return d.dht.Close()
+}
+
+func (d *discovery) discoverAndAdvertise() error {
+	rd := libp2pdiscovery.NewRoutingDiscovery(d.dht)
+
+	err := d.dht.Bootstrap(d.ctx)
+	if err != nil {
+		return fmt.Errorf("failed to bootstrap DHT: %w", err)
+	}
+
+	// wait to connect to bootstrap peers
+	time.Sleep(time.Second)
+	peersToTry := make(map[*peer.AddrInfo]struct{})
+
+	go func() {
+		ttl := time.Second * 30
+
+		for {
+			select {
+			case <-time.After(ttl):
+				logger.Info("advertising ourselves in the DHT...")
+				err := d.dht.Bootstrap(d.ctx)
+				if err != nil {
+					logger.Warn("failed to bootstrap DHT", "error", err)
+					continue
+				}
+
+				ttl, err = rd.Advertise(d.ctx, string(d.pid))
+				if err != nil {
+					logger.Warn("failed to advertise in the DHT", "error", err)
+					ttl = time.Minute
+				}
+			case <-d.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	go func() {
+		logger.Info("attempting to find peers...")
+		peerCh, err := rd.FindPeers(d.ctx, string(d.pid))
+		if err != nil {
+			logger.Error("failed to begin finding peers via DHT", "err", err)
+			return
+		}
+
+		for {
+			select {
+			case <-d.ctx.Done():
+				return
+			case <-time.After(time.Minute):
+				if len(d.h.Network().Peers()) > d.minPeers {
+					continue
+				}
+
+				// reconnect to peers if peer count is low
+				for p := range peersToTry {
+					logger.Info("trying to connect to cached peer", "peer", p.ID)
+					err = d.h.Connect(d.ctx, *p)
+					if err != nil {
+						logger.Info("failed to connect to discovered peer", "peer", p.ID, "err", err)
+					}
+				}
+			case peer := <-peerCh:
+				if peer.ID == d.h.ID() || peer.ID == "" {
+					continue
+				}
+
+				logger.Info("found new peer via DHT", "peer", peer.ID)
+
+				// found a peer, try to connect if we need more peers
+				if len(d.h.Network().Peers()) < d.maxPeers {
+					err = d.h.Connect(d.ctx, peer)
+					if err != nil {
+						logger.Info("failed to connect to discovered peer", "peer", peer.ID, "err", err)
+					}
+				} else {
+					d.h.Peerstore().AddAddrs(peer.ID, peer.Addrs, peerstore.PermanentAddrTTL)
+					peersToTry[&peer] = struct{}{}
```

Comment:
maybe will we need to check if the peer that arrived is already on the list?

Reply:
Will be a good idea if we have a limit to add new peers on the `peersToTry` list? then we avoid a long list, and when the list is already full, we test the connection of the first item (most aged peer) and if the connection fails, then we remove it and add the new peer to the end of the list (newest peer)?

Reply:
for the first comment, the map ensures no duplicates. and yes I'm not 100% sure about using the peersToTry list yet or just using the peerstore, either way I'm going to clean it up before opening the PR